code/tsp_ls2.py

```diff
@@ -2,12 +2,152 @@
+    total_cost += initialdist_matrix[route[len(route)-1],route[0]].cost
+    route.append(route[0])
+    return route,total_cost    
+
+def accep_criteria(neighbor_route,current_route,cost_matrix,temperature):
+    neighbor_route_cost = calculate_route_cost(neighbor_route,cost_matrix)
+    current_route_cost = calculate_route_cost(current_route,cost_matrix)
+    if neighbor_route_cost < current_route_cost:
+        return neighbor_route,calculate_route_cost(neighbor_route,cost_matrix)
+    else:
+        probability = math.exp(-(neighbor_route_cost - current_route_cost) / temperature)
+        if probability >= random.random():
+            return neighbor_route,calculate_route_cost(neighbor_route,cost_matrix)
+    return current_route,current_route_cost
+    
+def anneal_route(nodes,start_nodes,initial_T,ending_T,cost_matrix, tracepoint_pipe : Pipe, solution_pipe : Pipe):
+    start_time = time.process_time()
+    a = .995 #cooling constant
+    mat = distance_calculator(nodes)
+    initial = 1
+    while 1:
+        for x in start_nodes:
+            temperature = initial_T
+            greedy_route,greedy_total_cost = greedy_heuristic(mat,x)
+            if initial == 1:
+                best_route = greedy_route
+                best_cost = greedy_total_cost
+                initial = 0
+            num_cities = len(greedy_route)-1
+            current_route = greedy_route
+            while temperature >= ending_T:
+                adder_rand = random.randint(2,num_cities-1)
+                left_rand = random.randint(1,num_cities-adder_rand)
+                neighbor_route = current_route[:]
+                neighbor_route[left_rand:(left_rand + adder_rand)] = reversed(neighbor_route[left_rand:(left_rand + adder_rand)])
+                current_route,current_cost = accep_criteria(neighbor_route,current_route,cost_matrix,temperature)
+                if current_cost < best_cost:
+                    best_cost = current_cost
+                    best_route = current_route
+                    print("Optimal Updated by Annealing")
+                    solution_pipe.send(Solution( best_cost, best_route))
+                    tracepoint_pipe.send( Tracepoint( time.process_time() - start_time, best_cost ) )
+                if  current_cost-best_cost > best_cost*.5:
+                    print("Resetting Annealing")
+                    current_route = greedy_route
+                    current_cost = greedy_total_cost
+                    temperature = initial_T
+                temperature = temperature*a
+    solution_pipe.send(Solution( best_cost, best_route))
+    tracepoint_pipe.send( Tracepoint( time.process_time() - start_time, best_cost ) )
+    
 
 def ls2( nodes , timeout : int,seed_num ): #Here we will implement the Simulated Annealing Algorithm
+    start_time = time.time()# Get current uptime in secconds
+    solution_read, solution_write = Pipe()
+    tracepoint_read, tracepoint_write = Pipe()
+    random.seed(seed_num)
```

Comment:
Perhaps lines 114 through 122 (excluding line 117) should be moved to the anneal_route function, since they are doing algorithm work?

In a similar vein, maybe it also makes sense for each algorithm to construct the list of nodes.

Reply:
yeah that makes sense im going to move the list of node construction into the algorithms